Smarter Resources

swfarnsworth opened the pull request:
This PR contains the first iteration of the new resources page. I expect that a few of the tests will not pass but am otherwise ready for criticism to prepare it for production.

Commit:
Add bulma-modal, we'll need it later.

Commit:
Dramatically simplify resources.

We don't need _category_info, we don't need subcategories, we this will
be much simpler now.

Also, rglob is nice.

Commit:
Add GoodReads CSS.

Commit:
Add tags to all the resources.

Also updates all amazon links with GoodReads.

Commit:
Remove www from the URL for inferential thinking.

The www subdomain gives an SSL error for an expired certificate, so we
don't want to use that one.

Commit:
Test boilerplate

swfarnsworth committed:
Collapse resources into one directory, migrate to new tagging system.

This commit fails the lint test from pipenv, though the main branch is on poetry now.

lemonsaurus committed:
Update templates with new resource urls.

swfarnsworth committed:
Merge branch 'main' of https://github.com/python-discord/site into swfarnsworth/smarter-resources/new-resources

swfarnsworth committed:
Add final line break for style compliance.

swfarnsworth committed:
Add resource Data Science from Scratch.

swfarnsworth committed:
Add resource regex101.

swfarnsworth committed:
Remove reference to Python 2.7.

The maintainer of regex101 has changed the name from "Python 2.7" to just "Python".

swfarnsworth committed:
Add Kaggle's pandas tutorial.

fisher60 committed:
change resources to prepare for smart resource search

fisher60 committed:
add function to gather tags from resource yaml files

I added a function to utils to load all the tags from the resources, this should allow all the tags to be loaded once and exist in memory.

fisher60 committed:
add constant for resource tags. Add Functional front end

This change adds a fully functional front end menu for testing. This is a very rough outline and will need a bit of UI and UX love to get working fully. Should act as an example of functionality

swfarnsworth committed:
Remove extra newline from end of file (there is now one newline).

swfarnsworth committed:
Added type annotations; refactored.

swfarnsworth committed:
Updated type annotations; get_resources_meta now returns a dict of sorted lists of strings in title-case.

fisher60 committed:
working demo of smart resources search

I have incorporated a search that allows users to check boxes to filter resources. This is a working version, but the algo for searching likely needs to be modified. The frontend also needs some style updates. All necessary functionality should be present now though.

fisher60 committed:
refactor search function for resources

fisher60 committed:
remove redundant variable declaration

swfarnsworth committed:
Create Resource type annotation.

swfarnsworth committed:
Restructure table of resources for new query logic.

fisher60 committed:
split up checkbox options in query params

I have added functionality to the front end to split up the checkbox options into separate query params. This will make it easier to parse what has been selected.

swfarnsworth committed:
Convert data structure for query.

Was list[str], is now dict[str, set[str]], organized by category.

swfarnsworth committed:
Implement desired query logic.

swfarnsworth committed:
Add empty set as an initial value.

Otherwise, the call to reduce will raise an error when no options for a given category are selected.

swfarnsworth committed:
Sort the columns in the UI.

swfarnsworth committed:
Delete unused `yaml_file_matches_search` method.

swfarnsworth committed:
"python" -> "general" in tags/topics.

As most of the resources are Python related, tagging them with "python" is uninformative.

Commit:
styles: filtering resources page

doublevcodes committed:
feat: Make default options ticked when page loads

doublevcodes committed:
chore: Update django-simple-bulma from v2.1 to v2.4

This is needed because the flexbox helper classes were only introduced in later versions (Bulma v0.9.1 and djano-simple-bulma v2.4.0)

doublevcodes committed:
fix: Pin `psycopg2-binary` to v2.8.*

`psycopg2-binary` introduced a change that cause Django to break in v2.9. Without this pin, starting up the site raises an `AssertionError: database connection isn't set to UTC`

doublevcodes committed:
feat: Center `Search` and `Clear Search` buttons

doublevcodes committed:
feat: Add a source button to the resource boxes

This button should be used by maintainers to take them to the YAML file holding data about that resource. It currently has no functionality.

doublevcodes committed:
lint: Add a NOQA for the N818 linting error

This newly added rule enforces exception names to end with Error or Exception, causing the code to not pass linting.

I preferred ignoring this error instead of fixing it to keep the style consistent with rest_framework.NotFound.

doublevcodes committed:
chore: Split Topic column into two columns

doublevcodes committed:
chore: Add tag to resource boxes

doublevcodes committed:
chore: Finalise tag style

doublevcodes committed:
feat: Add select all button

doublevcodes committed:
fix: Make select all buttons for each individual column

doublevcodes committed:
Align select all buttons to center of column

swfarnsworth committed:
New Resource: vcokltfre's Discord bot tutorial.

swfarnsworth committed:
New Resource: vcokltfre's Discord bot tutorial.

doublevcodes committed:
merge: Merge branch 'doublevcodes/smarter-resources/front-end' of https://github.com/doublevcodes/site into doublevcodes/smarter-resources/front-end

swfarnsworth committed:
Merge branch 'doublevcodes/smarter-resources/front-end' into swfarnsworth/smarter-resources/new-resources

swfarnsworth committed:
Make video resource boxes consistent.

The title now links to the YouTube channels rather than a YouTube icon. Channels with an associated Discord server are also tagged "community".

swfarnsworth committed:
Add nedbat's kindling projects.

This is the first "project ideas" resource.

swfarnsworth committed:
Make icon dict a constant; add project ideas icon.

swfarnsworth committed:
Use MappingProxyType to freeze non-private dictionaries.

This is intended to make those mappings safer.

swfarnsworth committed:
Make title for all resources a link.

When the resource is for a Discord server, the title links to a website and the Discord icon is the server invite.

swfarnsworth committed:
Rename `utils.py` to `resource_search.py`.

swfarnsworth committed:
Drop -Option from {topic,type,etc}Option.

Having -Option in the names was making the URL especially verbose.

swfarnsworth committed:
Make an asterisk in the URL a wildcard.

This means that /resources/?topic=*&type=Book&payment=Paid&complexity=*` would return paid books of all topics and complexities.

swfarnsworth committed:
Delete `resources_resources_redirect`.

We may need to redirect all pages that are being consolidated to the one resources page.

swfarnsworth committed:
Add `tutorial` and `video` tags for RealPython.

swfarnsworth committed:
Merge with main, resolve conflicts.

swfarnsworth commented on pydis_site/apps/redirect/redirects.yaml:
```diff
@@ -87,10 +87,6 @@ resources_index_redirect:
-resources_resources_redirect:
-  original_path: pages/resources/<str:category>/
-  redirect_route: "resources:resources"
-
```
There will be other missing pages now that we've consolidated the resources into one page, though I'm not sure how to create redirects for these.

whispersofthedawn commented on pydis_site/static/css/resources/resources.css:
```diff
@@ -1,29 +1,29 @@
-    padding: 0.65rem 0.85rem 0.85rem 0.65rem;
-    filter: saturate(1.1) brightness(1.1);
-}
+/*.box, .tile.is-parent {*/
```
Is this file going to be kept? It seems to be fully commented out.

whispersofthedawn commented on pydis_site/templates/resources/resources.html:
```diff
@@ -1,90 +1,199 @@
+                                </div>
+                            </div>
+                            <span class="control ml-2 is-flex is-align-items-end is-justify-content-center mt-auto">
+                                <button onclick="selectAllQueryParams('topic')" class="button is-success is-small">Select All</button>
```
This button does not seem to do anything if you click it when everything is already selected, is that intended?

whispersofthedawn reviewed (commented):
This looks pretty good to me. I don't know much Django so I may have missed some important things, but what I know looks fine.

brad90four commented on pydis_site/apps/resources/tests/test_views.py:
```diff
@@ -19,16 +19,13 @@ def test_resources_index_200(self):
+    @patch("pydis_site.apps.resources.resource_search.RESOURCES_PATH", TESTING_RESOURCES_PATH)
+    def test_filter_resource_list(self):
+        """TODO: Check that we can correctly filter resources with GET parameters."""
+        pass
```
So this test is just going to be skipped?

brad90four commented on pydis_site/apps/resources/tests/test_resources.py:
```diff
@@ -0,0 +1,9 @@
+
+    def test_utils_to_retrieve_tags(self):
+        """Test that the utils that retrieve the tags work as intended."""
+        pass
```
Skipping this test?

jchristgit replied:
@lemonsaurus what should we do here?

lemonsaurus replied:
We should just delete this file. There are no utils to retrieve tags anymore.

swfarnsworth committed:
Removed default selections.

We will instead treat empty selections as all being selected.

swfarnsworth committed:
Empty columns treated as all selected.

Treating an asterisk as a wild card is now redundant.

swfarnsworth committed:
Remove wildcard asterisks in URL.

This is no longer necessary as empty columns are treated as all being selected.

swfarnsworth committed:
Merge branch 'main' into swfarnsworth/smarter-resources/merge-with-main

lemonsaurus committed:
Kaizen: Refactor some code to be more beautiful.

lemonsaurus committed:
Always add a link icon for the primary URL.

lemonsaurus committed:
Move evil <script> tag into a .js file.

lemonsaurus committed:
Make GoodReads icons readable.

lemonsaurus committed:
Link icons belong close together.

lemonsaurus committed:
Merge branch 'main' into swfarnsworth/smarter-resources/merge-with-main

lemonsaurus committed:
Fix broken yaml in kivy.yaml.

lemonsaurus committed:
Greatly simplify the backend.

Here we're getting rid of all filtering and search functionality
on the backend. We'll be handling this on the client-side from now on.

lemonsaurus committed:
Add a template tag for css-classifying things.

lemonsaurus committed:
Add a FontAwesome token that we actually control.

Previously we were using some sort of borrowed token, but it
seems to not be working anymore for local sessions, so let's
just get rid of it.

lemonsaurus committed:
Clean up dependency file and bump simple-bulma.

lemonsaurus committed:
Resource filtering on the client, pt 1.

Here's the initial version of this system. We've got filtering,
but only by clicking checkboxes. The overall look and style are
pretty close to where we want them, but it's missing tons of
polish to be complete. The following commits will contain that polish.

lemonsaurus committed:
Reflect and retain filter state in URL params.

This commits makes two important changes:
- Whenever we click on filter checkboxes, the URL is immediately updated
  to include parameters that represent the chosen options. For example,
  if you select Beginner complexity, the url is updated with
  ?complexity=beginner.
- Whenever the page is loaded, we deserialize any parameters that exist
  in the URL and change the filter state to match it. In other words,
  refreshing the page at any time retains the current state,
  and you can now link people to exactly the filter results that you
  want to share.

lemonsaurus committed:
Fix broken icon for Paid resources.

lemonsaurus committed:
Make all tags interactive.

Adds a bunch of tag-related quality of life improvements:
- You can now see which filters you've added at the top left.
- You can click on tags to start filtering to that tag.
- You can click on tags in the filter box to remove the filter.
- Tags will now show an outline when they are active.

lemonsaurus committed:
Let the user know when there are no matches.

lemonsaurus committed:
Distill the resources page.

This should allow Netlify Deploy Previews.

lemonsaurus committed:
Don't round edges of the filter category headers.

lemonsaurus committed:
Use fullwidth icons so everything is aligned.

lemonsaurus committed:
Add Guides to the More menu.

Since we've removed the resources landing page, there's no
easy way to get to Guides anymore. For that reason, I'm adding it to
the More menu so that it'll still be discoverable.

lemonsaurus committed:
Complexity -> Difficulty.

lemonsaurus committed:
Sort all the resources alphabetically.

lemonsaurus committed:
Add redirects for all old resource endpoints.

This will maintain backwards compatibility once this is merged.

lemonsaurus committed:
Redirects from old endpoints now filter correctly.

For example, navigating to pydis.com/resources/communities will now
correctly redirect to pydis.com/resources/?type=community.

lemonsaurus committed:
Don't redirect if the resource_type is None!

lemonsaurus committed:
Clean up hover effects.

- We now have a hover effect for all icons.
- We also now support hover effects for all tags.
- However, we don't have any hover effect for icons in titles.
  That just looks weird.

Additionally, we're getting rid of some bloat and consolidating the CSS
into one place.

lemonsaurus committed:
Support dashful redirects.

Previously, trying to go to `resources/project%20ideas` would crash
the filtering JS.  This has now been sorted out, so that these types
of redirects have their spaces replaced by dashes, which makes them
valid again.

lemonsaurus committed:
Replace is-goodreads-cream with is-black.

The goodreads cream, while nice, has poor readability on a light
background. So I'm ditching it and just going black with those icons.

lemonsaurus committed:
Add plural redirects, like /books.

This is an intuitive thing to link to, so let's support it for better UX

lemonsaurus committed:
Ensure "Other" is always the bottom topic.

lemonsaurus committed:
Show a duck pond when no filters match.

lemonsaurus committed:
Fix resource redirects for deploy previews.

This could probably be improved by making it less hardcoded, but since
there's a bunch of special cases, and since it only affects deploy
previews, I'm choosing to solve this with a simple solution.

I don't think realistically that we're gonna be adding a ton of resource
types down the line anyway.

lemonsaurus committed:
Get rid of the title_url for university of toronto

lemonsaurus committed:
Fix broken css for no-resource duck pond.

lemonsaurus committed:
Show a "No filters selected" indicator.

lemonsaurus committed:
Remove italics from resource cards.

lemonsaurus committed:
Refactor: collapsibles as a stand-alone component

lemonsaurus committed:
Hide the filter tags initially.

Helps prevent a flash of tags.

lemonsaurus committed:
Make the category headers smaller on mobile.

lemonsaurus committed:
Collapse category headers on load for mobile only.

lemonsaurus committed:
Kaizen: Clean up some indentation.

lemonsaurus committed:
Revert ugly deploy preview redirect hack.

HassanAbouelela committed:
Merge branch 'main' into swfarnsworth/smarter-resources/merge-with-main

Signed-off-by: Hassan Abouelela <hassan@hassanamr.com>

# Conflicts:
#	poetry.lock
#	pyproject.toml

lemonsaurus committed:
Shrink and center the duck pond 404.

lemonsaurus committed:
Make a few minor corrections to resources.

- Changes Google Collab to Colab
- Updates the URL for the vcokltfre tutorial
- Adds a category icon for Game Development.

lemonsaurus committed:
Make the resource link icons a size smaller.

We also add a little breathing room to the tag container, so that it
doesn't encroach on the personal space of the icons.

lemonsaurus committed:
Indent checkboxes under the headers.

lemonsaurus committed:
Added margins and max-width to filtering column.

lemonsaurus committed:
Get rid of invalid filters in the URL.

lemonsaurus committed:
Hide no filters selected in all cases.

lemonsaurus committed:
Add Socratica to the resource list.

Closes https://github.com/python-discord/meta/issues/119

lemonsaurus committed:
Minor changes to socratica resource.

lemonsaurus committed:
Improve performance of addFilter() up to 300%

lemonsaurus committed:
Fix category header double click bug on mobile.

lemonsaurus committed:
Placate the flake8.

lemonsaurus committed:
Adding missing EOFs.

lemonsaurus committed:
Add a button for removing all active filters.

lemonsaurus committed:
Revert changes to pyproject.toml.

These changes were not necessary. I am very sorry for the things
that I did.

jchristgit commented on pydis_site/apps/redirect/redirects.yaml:
```diff
@@ -87,9 +87,45 @@ resources_index_redirect:
+resources_books_redirect:
+  original_path: resources/books/
+  redirect_route: "resources:index"
+  redirect_arguments: [ "book" ]
```
This should match the format of the other blocks.

```suggestion
  redirect_arguments: ["book"]
```

lemonsaurus committed:
Add more tests for the new resources page.

Coverage is now 100%.

jchristgit commented on pydis_site/apps/resources/views/resources.py:
```diff
@@ -1,7 +1,119 @@
+        if resource_type:
+            dashless_resource_type = resource_type.replace("-", " ")
+
+            if dashless_resource_type.title() not in self.filters['Type']['filters']:
```
```suggestion
            if dashless_resource_type.title() not in self.filters["Type"]["filters"]:
```

jchristgit commented on pydis_site/templates/resources/resource_box.html:
```diff
@@ -9,14 +11,69 @@
-        </span>
-    {% endfor %}
+    <div class="is-flex is-align-items-center">
+        {#  Add primary link #}
```
```suggestion
        {# Add primary link #}
```

jchristgit commented on pydis_site/templates/resources/resource_box.html:
```diff
@@ -9,14 +11,69 @@
-    {% endfor %}
+    <div class="is-flex is-align-items-center">
+        {#  Add primary link #}
+        {% if 'title_url' in resource %}
```
```suggestion
        {% if "title_url" in resource %}
```

jchristgit commented on pydis_site/templates/resources/resource_box.html:
```diff
@@ -9,14 +11,69 @@
+                <i class="{{ tag|title|get_category_icon }} mr-1"></i>
+                {{ tag|title }}
+            </span>
+            {% endfor %}
```
I am sort of wondering whether we can merge these together. The only thing that changes is `data-filter-name` and the thing we are iterating over, or am I misunderstanding something?

lemonsaurus replied:
they also all have different span classes (to give them different colors), so yeah we could do it, but we'd need to put those classes (like `has-background-info-light` and `has-text-info-dark`) into the dict we're iterating or something... Would be nice and clean, though.

lemonsaurus replied:
Refactoring this turns out to be quite a big job, and I'm not sure I think it's a good idea to start storing lots of CSS classes in the view logic. #wontfix.

jchristgit commented on pydis_site/templates/resources/resources.html:
```diff
@@ -1,90 +1,164 @@
+                                                        {{ filter_item|title }}
+                                                        <button class="delete is-small is-primary has-background-primary-light"></button>
+                                                    </span>
+                                            {% endif %}
```
Can we maybe simplify this a bit? Maybe move the `if` condition to the `button` itself to just control its tags? Or am I missing other diffs here?

lemonsaurus replied:
yeah so here the colors are both in the span class (`is-primary`) and in the delete button (`has-background-primary-light`), so you'd need twice as many if's, which wouldn't really be that much simpler. The only way to simplify it down to a single span would be to store all these colors somewhere in the view and make sure I can get them from filters.items, which adds more clutter to the view.. but might be a better solution. I'm not sure.

lemonsaurus replied:
Same thing here - I don't wanna store these color classes in the view. It violates the [locality of reference principle](https://en.wikipedia.org/wiki/Locality_of_reference), which feels like too great a cost to pay for cleaner template code.

The classes belong with the HTML element, not in some dict in the Python view.

lemonsaurus committed:
Make resource tags correctly deactivate.

lemonsaurus committed:
When hovering titles, make them light, not black.

lemonsaurus committed:
Fix bug where transition wouldn't work on first collapse.

lemonsaurus committed:
Sort resources alphabetically, disregarding 'the'.

lemonsaurus committed:
Fix delay when collapsing collapsibles.

lemonsaurus committed:
Add type annotations for sort key method.

lemonsaurus commented on pydis_site/apps/resources/templatetags/as_css_class.py:
```diff
@@ -0,0 +1,18 @@
+    For example, convert
+    "Favorite FROOT_is_LEMON" to
+    "favorite-froot-is-lemon"
+    """
```
This is called `kebab-case`, I learned today. Let's rename this function to use the correct term, and maybe cover more edge cases, too.

lemonsaurus commented on pydis_site/apps/resources/views/resources.py:
```diff
@@ -1,7 +1,128 @@
+                    css_tag = f"{tag_type}-{tag}"
+                    css_tag = css_tag.replace("_", "-")
+                    css_tag = css_tag.replace(" ", "-")
+                    css_classes.append(css_tag)
```
Why aren't we just using the as_css_class template tag here? This is just duplicating code from there.

lemonsaurus committed:
Kaizen: Make jchrists wildest dreams come true

Co-authored-by: Johannes Christ <jc@jchri.st>

lemonsaurus committed:
Murder some space

Remove those pointless empty pages at the
start and the end of a book.

Co-authored-by: Johannes Christ <jc@jchri.st>

lemonsaurus committed:
Refactor as_css_class to 'to_kebabcase'.

We're making a few changes here
- Changing the name to 'to_kebabcase'
- Covering all edge cases
- Adding a unit test for this utility

lemonsaurus committed:
Complete refactor of collapsibles.

This is now a completely self-contained feature, which
can be used in the same way on every page. I've moved all the
collapsible-related logics out of the resources.js file and
into collapsibles.js, and added documentation and other
quality-of-life features that will apply to other pages, as well.

Changes:
- The icon will now always change when the collapsible opens or closes.
- By adding the "collapsed" class, you can tell the collapsible to
  be collapsed by default.

lemonsaurus reviewed (approved):
I've addressed all the issues I was able to find with my own code.

jchristgit reviewed (approved):
Thank you lemon!

![image](https://user-images.githubusercontent.com/22679924/152061545-48a18226-d678-4280-bbc1-18b1a5775f74.png)

lemonsaurus committed:
Use external-link-alt for all links.

lemonsaurus committed:
Duck pond removed when removing all filters.

mbaruh reviewed (approved):
By the power vested in me, this LGTM
